worker/src/task.rs

```diff
@@ -0,0 +1,600 @@
+                );
+            }
+            task_id = state.next_task_id;
+            detached = state.task_tracker_state != TaskTrackerState::Running;
+            if detached {
+                warn!("Task arrived after task tracking is shutdown, will start as detached");
+            }
+            state.next_task_id = task_id + 1;
+        }
+        let shutdown_token = Shutdown::new();
+        let cloned_token = shutdown_token.clone();
+        let (tx, rx) = tokio::sync::oneshot::channel::<T::Output>();
+        let wrapped = shutdown_token.clone().wrap_cancel(run);
+        let fut = Box::pin(async move {
+            info!("Starting task with task_id {task_id}");
+            let task_output = wrapped.await;
+            if task_output.is_none() {
+                info!("Task with task_id {task_id} is cancelled because of shutdown");
+                drop(tx);
+                return ();
+            }
+            if tx.send(task_output.unwrap()).is_err() {
+                warn!("Task with task_id {task_id} output could not be sent: receiver dropped")
+            }
+            ()
+        });
+        self.get_task_handle::<T>(tag, task_id, cloned_token, fut, detached, rx)
+    }
+    pub fn spawn_wait<T>(&self, tag: Option<String>, run: T) -> TaskHandle<T::Output>
+    where
+        T: Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        let task_id;
+        let detached;
+        {
+            let mut state = self.state.lock();
+            if state.tasks.contains_key(&state.next_task_id) {
+                panic!(
+                    "Task tracker already running a task with next id: `{{state.next_task_id}}`"
+                );
+            }
+            task_id = state.next_task_id;
+            detached = state.task_tracker_state != TaskTrackerState::Running;
+            if detached {
+                warn!("Task arrived after task tracking is shutdown, will start as detached");
+            }
+            state.next_task_id = task_id + 1;
+        }
+        let shutdown_token = Shutdown::new();
+        let cloned_token = shutdown_token.clone();
+        let (tx, rx) = tokio::sync::oneshot::channel::<T::Output>();
+        let wrapped = shutdown_token.clone().wrap_wait(run);
+        let fut = Box::pin(async move {
+            info!("Starting task with task_id {task_id}");
+            if wrapped.is_err() {
+                error!("Task with task_id {task_id} is already shutdown before shutdown closure could run");
+                drop(tx);
+                return ();
+            }
+            let task_output = wrapped.unwrap().await;
+            if tx.send(task_output).is_err() {
+                warn!("Task with task_id {task_id} output could not be sent: receiver dropped")
+            }
+            ()
+        });
+        self.get_task_handle::<T>(tag, task_id, cloned_token, fut, detached, rx)
+    }
+    pub fn spawn_cancel_and_wait<T, U>(
```

Comment:
So if I understand correctly, the clean up here will be executed as a separate future - not related to the one we provided. For the components we use , ex the [block_waiter](https://github.com/MystenLabs/narwhal/blob/c0eafd241c368b7cba73bec9d350a037068c755e/primary/src/block_waiter.rs#L256) , we might need to perform clean ups from inside the component (object) it self. What would be your suggestion in this case? Implementing the drop method for the object to act as a clean up?

Reply:
I think for such a scenario, using Tasktracker::spawn(tag, closure, shutdown_token) will be ideal. User can pass int he token and use it inside the loop like in my test examples below